examples/pipelines/commoncrawl/components/download_commoncrawl_segments/src/main.py

```diff
@@ -0,0 +1,139 @@
+) -> List[List[str]]:
+    """Downloads a WARC file and extracts the webpages.
+    Args:
+        warc_file: The path to the WARC file.
+        use_s3: Whether to download the WARC file from S3 or from the Commoncrawl API.
+        get_plain_text: Whether to convert the HTML content to plain text.
+        n_records_to_download: The number of webpages to download from the WARC file.
+    Returns:
+        A list of webpages.
+    """
+    logger.info(f"Processing WARC file from segment path: {warc_file}...")
+
+    if use_s3:
+        response = get_warc_file_using_boto3(warc_file)
+        with gzip.GzipFile(fileobj=response, mode="rb") as file:
+            return get_records(file, get_plain_text, n_records_to_download)
+    else:
+        response = get_warc_file_using_requests(warc_file, retries, backoff_factor)
+        return get_records(response.raw, get_plain_text, n_records_to_download)
+
+
+class DownloadCommoncrawlSegments(DaskTransformComponent):
+    def __init__(
+        self,
+        *_,
+        use_s3: Optional[bool] = False,
+        get_plain_text: Optional[bool] = False,
+        n_records_to_download: Optional[int] = None,
+        retries: Optional[int] = None,
+        backoff_factor: Optional[float] = None,
+    ):
+        """Downloads Commoncrawl segments based on a list of WARC paths.
+        Args:
+            use_s3: Whether to download the WARC files from S3 or from the Commoncrawl API.
+            get_plain_text: Whether to convert the HTML content to plain text.
+            n_records_to_download: The number of webpages to download from each segment.
+        """
+        self.use_s3 = use_s3
+        self.get_plain_text = get_plain_text
+        self.n_records_to_download = n_records_to_download
+        self.retries = retries
+        self.backoff_factor = backoff_factor
+
+    def transform(
+        self,
+        dataframe: dd.DataFrame,
+    ) -> dd.DataFrame:
+        """Downloads Commoncrawl segments based on a list of WARC paths.
+        Args:
+            dataframe: A Dask DataFrame containing a column of WARC paths.
+        Returns:
+            A Dask DataFrame containing the downloaded webpages.
+        """
+
+        dataframe = (
+            dataframe.apply(
+                lambda row: get_records_from_warc_file(
+                    row["segment_path"],
+                    self.use_s3,
+                    self.get_plain_text,
+                    self.n_records_to_download,
+                    self.retries,
+                    self.backoff_factor,
+                ),
+                axis=1,
+                meta=("object"),
+            )
+            .explode()
+            .apply(pd.Series, meta={0: "object", 1: "object"})
+        )
+
+        dataframe.columns = [
+            "webpage_url",
+            "webpage_html",
+        ]
+
+        dataframe = dataframe.reset_index(drop=True)
+
+        logger.info(f"Downloaded {len(dataframe)} webpages from Commoncrawl.")
```

Comment:
`len()` triggers a compute here. We might be able to use `.size` here instead, which is evaluated lazily, but not sure how we can combine it with logging. Also fine for me to just remove this.

Reply:
Ok, I've removed it. I'll look into `.size` for local testing